### PR TITLE
Meta Cloud API Whatsapp Integration

### PR DESCRIPTION
**Pull request by SmittieC**
### Technical Description
                                                                                  
Adds support for Meta Cloud API as a new WhatsApp messaging provider, enabling direct integration with the WhatsApp Business Platform without requiring a third-party intermediary like Turn.io.

This is phase 1 of the integration where only text messages are supported. Audio messages are coming in a followup PR.
                                                                                                                                                                                                                                                              
  **Key changes:**
                                                                                                                                                                                                                                                              
  - **New messaging provider type (`meta_cloud_api`)**: Registered as a `MessagingProviderType` with its own config form (`MetaCloudAPIMessagingConfigForm`) supporting `business_id`, `access_token`, `app_secret`, and `verify_token` fields
  - **`MetaCloudAPIService`**: New `MessagingService` implementation that sends text messages via the Meta Graph API (v25.0) and resolves phone number IDs from E.164 numbers using the WhatsApp Business Account Phone Number Management API
  - **Webhook endpoint** (`/channels/whatsapp/meta/incoming_message`): Handles both the Meta GET verification handshake and incoming POST message payloads. Verifies `X-Hub-Signature-256` HMAC signatures using the configured `app_secret`
  - **Phone number ID resolution**: During channel creation, the form validates the phone number against the WhatsApp Business Account and stores the resolved `phone_number_id` in `extra_data`. This ID is used for routing incoming webhooks and as the
  `from` identifier when sending messages
  - **`WhatsappChannel` refactor**: Extracted a `from_identifier` property that returns either `phone_number_id` (Meta Cloud API) or `number` (other providers), removing duplicated `from_number`/`to_number` logic across `send_text_to_user`,
  `send_voice_to_user`, and `send_file_to_user`
  - **Meta Cloud API reuses the Turn.io message format** (`MetaCloudAPIMessage = TurnWhatsappMessage`) since both use the same WhatsApp Business API payload structure

I'll write up some user docs for this as well

  ### Migrations

  - [x] The migrations are backwards compatible

  Migration `0043_alter_messagingprovider_type` adds `meta_cloud_api` to the `MessagingProviderType` choices. This is backwards compatible — existing code will simply not encounter the new type.

  ### Demo

  To set up:
  1. Create a Meta Cloud API messaging provider with your WhatsApp Business Account ID, System User Access Token, App Secret, and Webhook Verify Token
  2. Create a WhatsApp channel on an experiment using the new provider
  3. Configure the webhook URL in your Meta App dashboard pointing to `/channels/whatsapp/meta/incoming_message`
  4. Send a WhatsApp message to the configured business number

  ### Docs and Changelog
  - [x] This PR requires docs/changelog update


 
<details>
<summary>Walkthrough guide (AI guide)</summary>
## Overview

This PR adds Meta Cloud API as a new WhatsApp messaging provider, enabling direct integration with the WhatsApp Business Platform without a third-party intermediary like Turn.io. Phase 1 supports text messages only (audio coming in a follow-up PR).

**Size:** Large (781 additions, 27 deletions, 18 files changed)

**Key areas:** `apps/service_providers/` (provider type, config, service), `apps/channels/` (webhook, views, forms, tasks), `apps/chat/channels.py` (WhatsappChannel refactor)

## Architecture Impact

- **New provider type** plugged into existing `MessagingProviderType` → `form_cls` → `get_messaging_service` dispatch. Clean extension, no existing patterns broken.
- **Single global webhook** routed by `phone_number_id` (unlike Turn.io's per-experiment URLs). New routing pattern in the codebase.
- **Signature verification after channel lookup** — deliberate choice since `app_secret` lives in encrypted provider config.
- **`from_identifier` property** on WhatsappChannel abstracts Meta vs. others, consolidating duplicated `from_number` logic.

## Review Comments Summary

CodeRabbit flagged docstring coverage below 80% threshold. Author addressed review comments across multiple commits. Key iterations: webhook signature verification, verify_token hashing, class-based view refactor, module-level function extraction.

---

## Recommended Reading Order

### Step 1 of 8: Provider Type & Migration

**Files:** `apps/service_providers/models.py`, `apps/service_providers/migrations/0044_alter_messagingprovider_type.py`

**Why first:** Everything else depends on the new enum value and the `extra_data` field on `MessagingProvider`.

New enum value:

```python
class MessagingProviderType(models.TextChoices):
    twilio = "twilio", _("Twilio")
    turnio = "turnio", _("Turn.io")
    sureadhere = "sureadhere", _("SureAdhere")
    slack = "slack", _("Slack")
+   meta_cloud_api = "meta_cloud_api", _("Meta Cloud API (WhatsApp)")
```

Dispatch wiring:

```python
    @property
    def form_cls(self):
        ...
+           case MessagingProviderType.meta_cloud_api:
+               return forms.MetaCloudAPIMessagingConfigForm

    def get_messaging_service(self, config):
        ...
+           case MessagingProviderType.meta_cloud_api:
+               return messaging_service.MetaCloudAPIService(**config)
```

New field on `MessagingProvider`:

```python
class MessagingProvider(BaseTeamModel, ProviderMixin):
    type = models.CharField(max_length=255, choices=...)
    name = models.CharField(max_length=255)
    config = encrypt(models.JSONField(default=dict))
+   extra_data = models.JSONField(default=dict, blank=True)
```

Migration adds the enum choice and the `extra_data` JSONField. Backwards compatible.

**What to notice:**
- `extra_data` is a plain JSONField (not encrypted) — stores `verify_token_hash` for efficient DB lookups. Sensitive values stay in encrypted `config`.
- Migration is backwards compatible — existing code never encounters the new type.

**Architecture:** Follows the established provider extension pattern: add enum → wire `form_cls` → wire `get_messaging_service`. No new patterns at this layer.

---

### Step 2 of 8: Provider Config Form

**File:** `apps/service_providers/forms.py`

**Why here:** Defines what fields are collected when creating a Meta Cloud API provider, and how the verify token hash gets persisted.

```python
class MetaCloudAPIMessagingConfigForm(ObfuscatingMixin, ProviderTypeConfigForm):
    obfuscate_fields = ["access_token", "app_secret", "verify_token"]

    business_id = forms.CharField(label=_("WhatsApp Business Account ID"))
    access_token = forms.CharField(label=_("System User Access Token"))
    app_secret = forms.CharField(
        label=_("App Secret"),
        help_text=_("Used to verify incoming webhook signatures (X-Hub-Signature-256)."),
    )
    verify_token = forms.CharField(
        label=_("Webhook Verify Token"),
        help_text=_("Token used by Meta to verify the webhook URL. Must match the token configured in your Meta app."),
    )

    def save(self, instance):
        instance = super().save(instance)
        verify_token = self.cleaned_data["verify_token"]
        instance.extra_data = {
            "verify_token_hash": hashlib.sha256(verify_token.encode()).hexdigest(),
        }
        return instance
```

**What to notice:**
- 3 fields obfuscated in UI (`access_token`, `app_secret`, `verify_token`), but `business_id` is not — just an account identifier.
- `save()` hashes verify_token with SHA-256, stores in `extra_data`. Enables webhook verification endpoint to do a DB lookup by hash.
- `save()` overwrites `extra_data` entirely (not merging). Fine since verify_token_hash is the only extra_data for this provider type.

---

### Step 3 of 8: MetaCloudAPIService

**File:** `apps/service_providers/messaging_service.py`

**Why here:** Core domain logic — how we talk to Meta's Graph API. Phone number resolution and message sending live here.

Class definition:

```python
class MetaCloudAPIService(MessagingService):
    _type: ClassVar[str] = "meta_cloud_api"
    supported_platforms: ClassVar[list] = [ChannelPlatform.WHATSAPP]
    voice_replies_supported: ClassVar[bool] = False
    supported_message_types = [MESSAGE_TYPES.TEXT]

    access_token: str
    business_id: str
    app_secret: str = ""
    verify_token: str = ""
    _phone_number_id: str | None = pydantic.PrivateAttr(default=None)

    META_API_BASE_URL: ClassVar[str] = "https://graph.facebook.com/v25.0"
    META_API_TIMEOUT: ClassVar[int] = 30
    WHATSAPP_CHARACTER_LIMIT: ClassVar[int] = 4096
```

Phone number resolution:

```python
    def is_valid_number(self, number: str) -> bool:
        self._phone_number_id = self._fetch_phone_number_id(number)
        if not self._phone_number_id:
            raise ValueError(
                f"{number} was not found in the WhatsApp Business Account. "
                "Please verify the number is registered with your business."
            )
        return True

    def get_phone_number_id(self) -> str | None:
        return self._phone_number_id

    def _fetch_phone_number_id(self, phone_number: str) -> str | None:
        url = f"{self.META_API_BASE_URL}/{self.business_id}/phone_numbers"
        response = httpx.get(
            url, headers=self._headers,
            params={"fields": "id,display_phone_number"},
            timeout=self.META_API_TIMEOUT
        )
        response.raise_for_status()
        for entry in response.json().get("data", []):
            display = entry.get("display_phone_number", "")
            try:
                parsed = phonenumbers.parse(display)
                normalized = phonenumbers.format_number(parsed, phonenumbers.PhoneNumberFormat.E164)
            except phonenumbers.NumberParseException:
                continue
            if normalized == phone_number:
                return entry["id"]
        return None
```

Sending messages:

```python
    def send_text_message(self, message: str, from_: str, to: str, platform: ChannelPlatform, **kwargs):
        url = f"{self.META_API_BASE_URL}/{from_}/messages"
        chunks = smart_split(message, chars_per_string=self.WHATSAPP_CHARACTER_LIMIT)
        for chunk in chunks:
            data = {
                "messaging_product": "whatsapp",
                "to": to,
                "type": "text",
                "text": {"body": chunk},
            }
            response = httpx.post(url, headers=self._headers, json=data, timeout=self.META_API_TIMEOUT)
            response.raise_for_status()
```

**What to notice:**
- `is_valid_number` has a **side effect**: caches resolved `phone_number_id` in `_phone_number_id`. The form later calls `get_phone_number_id()` to retrieve it. Avoids a second API call but couples the two methods.
- `from_` in `send_text_message` is the `phone_number_id`, not a phone number. URL is `/{phone_number_id}/messages`.
- Message splitting via `smart_split` at 4096 chars — WhatsApp's per-message limit.
- Only `TEXT` in `supported_message_types` — voice/file support coming in phase 2.
- `_fetch_phone_number_id` fetches ALL phone numbers and iterates. Pagination may be needed for accounts with many numbers.

---

### Step 4 of 8: Channel Form & Data Model

**Files:** `apps/channels/forms.py`, `apps/channels/datamodels.py`

**Why here:** The form is where `phone_number_id` resolution happens at channel creation. The data model shows message format reuse.

Data model — simple type alias:

```python
# Meta Cloud API uses the same WhatsApp Business API message format as Turn.io
MetaCloudAPIMessage = TurnWhatsappMessage
```

Form refactor — `clean_number` split into `clean_number` (parse) + `clean` (validate):

```diff
 class WhatsappChannelForm(WebhookUrlFormBase):
     def clean_number(self):
         try:
             number_obj = phonenumbers.parse(self.cleaned_data["number"])
-            number = phonenumbers.format_number(number_obj, phonenumbers.PhoneNumberFormat.E164)
-            service = self.messaging_provider.get_messaging_service()
+            return phonenumbers.format_number(number_obj, phonenumbers.PhoneNumberFormat.E164)
+        except phonenumbers.NumberParseException:
+            raise forms.ValidationError("Enter a valid phone number (e.g. +12125552368).") from None
+
+    def clean(self):
+        cleaned_data = super().clean()
+        number = cleaned_data.get("number")
+        if not number or not self.messaging_provider:
+            return cleaned_data
+
+        service = self.messaging_provider.get_messaging_service()
+        try:
             if not service.is_valid_number(number):
                 self.warning_message = (
                     f"{number} was not found at the provider. Please make sure it is there before proceeding"
                 )
-            return number
-        except phonenumbers.NumberParseException:
-            raise forms.ValidationError("Enter a valid phone number (e.g. +12125552368).") from None
+        except ValueError as e:
+            self.add_error("number", str(e))
+            return cleaned_data
+        except Exception:
+            self.add_error("number", "Could not validate this number right now. Please try again.")
+            return cleaned_data
+
+        if self.messaging_provider.type == MessagingProviderType.meta_cloud_api:
+            cleaned_data["phone_number_id"] = service.get_phone_number_id()
+        return cleaned_data
```

**What to notice:**
- Refactoring from `clean_number` to `clean` is better Django practice — cross-field validation belongs in `clean()`.
- Meta's `is_valid_number` raises `ValueError` (hard error, no phone_number_id = can't function). Turn.io/Twilio return `False` (soft warning).
- `phone_number_id` flows into `cleaned_data` → `ChannelFormWrapper.save()` → `channel.extra_data` automatically.
- Generic `Exception` catch is a resilience measure — if the Meta API is down, the form gracefully errors.

---

### Step 5 of 8: Webhook Helpers

**File:** `apps/channels/meta_webhook.py` (NEW FILE)

**Why here:** Security-critical building blocks used by the view. Three pure module-level functions.

```python
def extract_message_values(data: dict) -> list[dict]:
    """Extract value dicts that contain messages from Meta webhook payload."""
    values = []
    for entry in data.get("entry", []):
        for change in entry.get("changes", []):
            value = change.get("value", {})
            if value.get("messages") and value.get("metadata", {}).get("phone_number_id"):
                values.append(value)
    return values


def verify_webhook(request) -> HttpResponse:
    """Handle the Meta webhook GET verification handshake."""
    mode = request.GET.get("hub.mode")
    token = request.GET.get("hub.verify_token")
    challenge = request.GET.get("hub.challenge")

    if mode != "subscribe" or not token or not challenge:
        return HttpResponseBadRequest("Verification failed.")

    token_hash = hashlib.sha256(token.encode()).hexdigest()
    exists = MessagingProvider.objects.filter(
        type=MessagingProviderType.meta_cloud_api,
        extra_data__verify_token_hash=token_hash,
    ).exists()

    if exists:
        return HttpResponse(challenge, content_type="text/plain")
    return HttpResponseBadRequest("Verification failed.")


def verify_signature(payload: bytes, signature_header: str, app_secret: str) -> bool:
    """Verify the X-Hub-Signature-256 header from Meta webhooks."""
    if not signature_header.startswith("sha256=") or not app_secret:
        return False

    expected_signature = signature_header.removeprefix("sha256=")
    computed = hmac.new(
        app_secret.encode(),
        payload,
        hashlib.sha256,
    ).hexdigest()
    return constant_time_compare(computed, expected_signature)
```

**What to notice:**
- **Hash-based DB lookup** for verify_token — avoids iterating all providers. Smart use of Django JSONField lookups.
- **`constant_time_compare`** prevents timing attacks on HMAC. Good security practice.
- `extract_message_values` filters for entries with both `messages` AND `phone_number_id` — skips delivery receipts (`statuses`) that come through the same webhook.
- All three are module-level functions — clean, testable, no state.

---

### Step 6 of 8: Webhook View & URL Routing

**Files:** `apps/channels/views.py`, `apps/channels/urls.py`, `apps/channels/models.py`

**Why here:** HTTP integration layer that ties everything together.

URL registration:

```python
path("whatsapp/meta/incoming_message",
     views.MetaCloudAPIWebhookView.as_view(),
     name="new_meta_cloud_api_message"),
```

Webhook URL property on `ExperimentChannel`:

```python
        elif provider_type == MessagingProviderType.meta_cloud_api:
            uri = reverse("channels:new_meta_cloud_api_message")
```

The view:

```python
@method_decorator(waf_allow(WafRule.NoUserAgent_HEADER), name="dispatch")
@method_decorator(csrf_exempt, name="dispatch")
class MetaCloudAPIWebhookView(View):
    def get(self, request):
        log.debug("Meta Cloud API webhook verification request received")
        return meta_webhook.verify_webhook(request)

    def post(self, request):
        try:
            data = json.loads(request.body)
        except json.JSONDecodeError:
            log.debug("Meta Cloud API webhook received invalid JSON")
            return HttpResponseBadRequest("Invalid JSON.")

        if data.get("object") != "whatsapp_business_account":
            log.debug("Meta Cloud API webhook ignored: object=%s", data.get("object"))
            return HttpResponse()

        try:
            message_values = meta_webhook.extract_message_values(data)
            if not message_values:
                log.debug("Meta Cloud API webhook received payload with no messages")
                return HttpResponse()

            first_phone_number_id = message_values[0]["metadata"]["phone_number_id"]
        except (KeyError, IndexError):
            log.debug("Meta Cloud API webhook payload missing expected fields")
            return HttpResponse()

        channel = (
            ExperimentChannel.objects.filter(
                platform=ChannelPlatform.WHATSAPP,
                extra_data__phone_number_id=first_phone_number_id,
                messaging_provider__type=MessagingProviderType.meta_cloud_api,
            )
            .select_related("experiment", "team", "messaging_provider")
            .first()
        )
        if not channel:
            log.info("Meta Cloud API webhook: no channel found for incoming payload")
            return HttpResponse()

        # Signature verification happens after channel lookup because the app_secret
        # needed to verify is stored in the channel's messaging provider config.
        app_secret = channel.messaging_provider.config.get("app_secret", "")
        signature = request.headers.get("X-Hub-Signature-256", "")
        if not meta_webhook.verify_signature(request.body, signature, app_secret):
            log.warning("Meta Cloud API webhook signature verification failed for channel %s", channel.id)
            return HttpResponse()

        set_current_team(channel.team)
        request.experiment = channel.experiment

        log.debug("Meta Cloud API webhook dispatching %d message(s) for channel %s", len(message_values), channel.id)
        for value in message_values:
            tasks.handle_meta_cloud_api_message.delay(
                channel_id=channel.id,
                team_slug=channel.team.slug,
                message_data=value,
            )

        return HttpResponse()
```

**What to notice:**
- **Single global endpoint** — no experiment ID in URL (unlike Turn.io). Routing by `phone_number_id` in payload.
- **Signature verification after channel lookup** — explicitly documented. `app_secret` is in encrypted provider config so we need the channel first.
- **All error paths return 200** — Meta retries on non-2xx and can disable webhooks. This is Meta's documented best practice.
- `csrf_exempt` + `waf_allow(NoUserAgent_HEADER)` — Meta webhooks have no CSRF tokens or standard user-agent headers.
- Good query optimization with `select_related`.

**Architecture:** New routing pattern. Turn.io uses per-experiment URLs (`/turn/<experiment_id>/`). Meta uses a single endpoint with payload-based routing. Both coexist cleanly.

---

### Step 7 of 8: WhatsappChannel Refactor & Task

**Files:** `apps/chat/channels.py`, `apps/channels/tasks.py`

**Why here:** Where message processing happens after webhook dispatches. The `from_identifier` refactor is the key DRY improvement.

New `from_identifier` property:

```python
    @property
    def from_identifier(self) -> str:
        """Returns the phone number ID for Meta Cloud API, or the phone number for other providers."""
        extra_data = self.experiment_channel.extra_data
        if self.experiment_channel.messaging_provider.type == MessagingProviderType.meta_cloud_api:
            phone_number_id = extra_data.get("phone_number_id")
            if not phone_number_id:
                raise ValueError("Meta Cloud API channel is missing phone_number_id in extra_data")
            return phone_number_id
        return extra_data["number"]
```

Simplified send methods (same pattern for all three):

```diff
     def send_text_to_user(self, text: str):
-        from_number = self.experiment_channel.extra_data["number"]
-        to_number = self.participant_identifier
         self.messaging_service.send_text_message(
-            message=text, from_=from_number, to=to_number, platform=ChannelPlatform.WHATSAPP
+            message=text, from_=self.from_identifier, to=self.participant_identifier, platform=ChannelPlatform.WHATSAPP
         )

     def send_voice_to_user(self, synthetic_voice: SynthesizedAudio):
-        from_number = self.experiment_channel.extra_data["number"]
-        to_number = self.participant_identifier
         self.messaging_service.send_voice_message(
-            synthetic_voice, from_=from_number, to=to_number, platform=ChannelPlatform.WHATSAPP
+            synthetic_voice, from_=self.from_identifier, to=self.participant_identifier, platform=ChannelPlatform.WHATSAPP,
         )

     def send_file_to_user(self, file: File):
-        from_number = self.experiment_channel.extra_data["number"]
-        to_number = self.participant_identifier
         self.messaging_service.send_file_to_user(
-            from_=from_number, to=to_number,
+            from_=self.from_identifier, to=self.participant_identifier,
             ...
         )
```

New Celery task:

```python
@shared_task(bind=True, base=TaskbadgerTask, ignore_result=True)
def handle_meta_cloud_api_message(self, channel_id: int, team_slug: str, message_data: dict):
    message = MetaCloudAPIMessage.parse(message_data)
    experiment_channel = (
        ExperimentChannel.objects.filter(
            id=channel_id,
            experiment__is_archived=False,
        )
        .select_related("experiment", "team", "messaging_provider")
        .first()
    )
    if not experiment_channel:
        log.info("No experiment channel found for channel_id=%s team=%s", channel_id, team_slug)
        return
    channel = WhatsappChannel(experiment_channel.experiment.default_version, experiment_channel)
    update_taskbadger_data(self, channel, message)
    channel.new_user_message(message)
```

Also added `messaging_provider` to `select_related` in shared `get_experiment_channel`:

```diff
-    return query.select_related("experiment", "team").first()
+    return query.select_related("experiment", "team", "messaging_provider").first()
```

**What to notice:**
- `from_identifier` eliminates duplication across 3 send methods. Future providers benefit from this abstraction too.
- Task takes `channel_id` directly — view already resolved the channel, avoiding redundant queries.
- Adding `messaging_provider` to `select_related` in `get_experiment_channel` benefits all providers.

---

### Step 8 of 8: Tests

**Files:** `apps/channels/tests/test_meta_cloud_api_webhook.py` (new, 242 lines), `apps/channels/tests/test_forms.py`, `apps/service_providers/tests/test_messaging_providers.py`

Webhook test fixtures:

```python
APP_SECRET = "test_app_secret"
VERIFY_TOKEN = "test_verify_token"

def _make_signature(payload: bytes, secret: str = APP_SECRET) -> str:
    sig = hmac.new(secret.encode(), payload, hashlib.sha256).hexdigest()
    return f"sha256={sig}"

@pytest.fixture()
def meta_cloud_api_provider():
    return MessagingProviderFactory.create(
        type=MessagingProviderType.meta_cloud_api,
        config={
            "access_token": "test_token",
            "business_id": "biz_123",
            "app_secret": APP_SECRET,
            "verify_token": VERIFY_TOKEN,
        },
        extra_data={
            "verify_token_hash": hashlib.sha256(VERIFY_TOKEN.encode()).hexdigest(),
        },
    )

@pytest.fixture()
def meta_cloud_api_channel(meta_cloud_api_provider):
    return ExperimentChannelFactory.create(
        platform=ChannelPlatform.WHATSAPP,
        messaging_provider=meta_cloud_api_provider,
        experiment__team=meta_cloud_api_provider.team,
        extra_data={"number": "+15551234567", "phone_number_id": "12345"},
    )
```

Signature verification tests:

```python
class TestMetaCloudAPIWebhookVerifySignature:
    def test_valid_signature(self):
        payload = b'{"test": "data"}'
        signature = _make_signature(payload)
        assert meta_webhook.verify_signature(payload, signature, APP_SECRET) is True

    def test_invalid_signature(self):
        assert meta_webhook.verify_signature(b'{"test": "data"}', "sha256=invalid", APP_SECRET) is False

    def test_missing_sha256_prefix(self):
        payload = b'{"test": "data"}'
        sig = hmac.new(APP_SECRET.encode(), payload, hashlib.sha256).hexdigest()
        assert meta_webhook.verify_signature(payload, sig, APP_SECRET) is False

    def test_empty_app_secret(self):
        payload = b'{"test": "data"}'
        signature = _make_signature(payload)
        assert meta_webhook.verify_signature(payload, signature, "") is False
```

POST flow tests:

```python
class TestNewMetaCloudApiMessage:
    def _post(self, payload_dict, app_secret=APP_SECRET):
        factory = RequestFactory()
        body = json.dumps(payload_dict).encode()
        signature = _make_signature(body, app_secret)
        request = factory.post("/", data=body, content_type="application/json",
                               HTTP_X_HUB_SIGNATURE_256=signature)
        return MetaCloudAPIWebhookView.as_view()(request)

    @patch("apps.channels.tasks.handle_meta_cloud_api_message.delay")
    def test_valid_message_returns_200(self, mock_delay, meta_cloud_api_channel):
        response = self._post(_meta_webhook_payload())
        assert response.status_code == 200
        mock_delay.assert_called_once()

    def test_invalid_signature_returns_200(self, meta_cloud_api_channel):
        """Invalid signature returns 200 to prevent Meta from retrying."""
        factory = RequestFactory()
        body = json.dumps(_meta_webhook_payload()).encode()
        request = factory.post("/", data=body, content_type="application/json",
                               HTTP_X_HUB_SIGNATURE_256="sha256=invalid")
        response = MetaCloudAPIWebhookView.as_view()(request)
        assert response.status_code == 200
```

Phone number ID resolution tests (5 parametrized cases):

```python
@pytest.mark.parametrize(
    ("api_data", "lookup_number", "expected_id"),
    [
        pytest.param(
            [{"id": "111", "display_phone_number": "+1 (212) 555-2368"},
             {"id": "222", "display_phone_number": "+27 81 234 5678"}],
            "+12125552368", "111", id="formatted_number",
        ),
        pytest.param(
            [{"id": "333", "display_phone_number": "+27812345678"}],
            "+27812345678", "333", id="e164_number",
        ),
        pytest.param(
            [{"id": "111", "display_phone_number": "+1 212 555 2368"}],
            "+27812345678", None, id="no_match",
        ),
        pytest.param([], "+12125552368", None, id="empty_response"),
        pytest.param(
            [{"id": "111", "display_phone_number": "not-a-number"},
             {"id": "222", "display_phone_number": "+27 81 234 5678"}],
            "+27812345678", "222", id="unparseable_number_skipped",
        ),
    ],
)
@patch("apps.service_providers.messaging_service.httpx.get")
def test_meta_cloud_api_get_phone_number_id(mock_get, meta_cloud_api_service, api_data, lookup_number, expected_id):
    mock_get.return_value = _mock_phone_numbers_response(api_data)
    assert meta_cloud_api_service._fetch_phone_number_id(lookup_number) == expected_id
```

Form tests:

```python
@patch("apps.channels.forms.ExtraFormBase.messaging_provider", new_callable=PropertyMock)
@patch("apps.service_providers.messaging_service.httpx.get")
def test_whatsapp_form_meta_cloud_api_resolves_phone_number_id(mock_httpx_get, messaging_provider, experiment):
    mock_httpx_get.return_value = httpx.Response(
        200, json={"data": [{"id": "12345", "display_phone_number": "+1 (212) 555-2368"}]},
        request=httpx.Request("GET", "https://test"),
    )
    provider = MessagingProviderFactory.create(
        type=MessagingProviderType.meta_cloud_api,
        config={"access_token": "test_token", "business_id": "biz_123"},
    )
    messaging_provider.return_value = provider
    form = WhatsappChannelForm(experiment=experiment, data={"number": "+12125552368", "messaging_provider": provider.id})
    assert form.is_valid()
    assert form.cleaned_data["phone_number_id"] == "12345"

def test_whatsapp_form_meta_cloud_api_rejects_unknown_number(...):
    # ... (empty response from API)
    assert not form.is_valid()
    assert "was not found in the WhatsApp Business Account" in form.errors["number"][0]
```

**What to notice:**
- Coverage is thorough for phase 1 — security paths, happy paths, and edge cases all covered.
- Tests verify invalid/missing signatures return 200 — matching Meta best practice.
- Also fixes a flaky test in `test_chatbot_tables.py` (clearing cached team slugs).

---

## Potential Concerns

- **`is_valid_number` side-effect coupling** — The validation method caches `phone_number_id` as a side effect. Could be clearer as a method that returns the ID directly.
- **No pagination for phone number listing** — `_fetch_phone_number_id` fetches all numbers without handling pagination. May fail for large business accounts.
- **Phase 2 gap** — `send_voice_message` and `send_file_to_user` not implemented. Calling them on a Meta channel would hit the base class.
- **Single channel per phone_number_id** — The webhook view uses `.first()` when looking up channels. If multiple channels share the same `phone_number_id`, only the first will receive messages.

</details>

### CODE REVIEW

File: apps/chat/channels.py [hunk 1]
```diff
@@ -1154,34 +1154,33 @@ def supports_multimedia(self) -> bool:
     def supported_message_types(self):
         return self.messaging_service.supported_message_types
 
+    @property
+    def from_identifier(self) -> str:
+        """Returns the phone number ID for Meta Cloud API, or the phone number for other providers."""
+        extra_data = self.experiment_channel.extra_data
+        return extra_data.get("phone_number_id") or extra_data["number"]
+
```
[the remark this answers is not in the document]
Comment (SmittieC) [reply apps/chat/channels.py]: Fixed in ee17335d1. `from_identifier` now checks `messaging_provider.type` and raises `ValueError` when `phone_number_id` is missing for Meta Cloud API channels, instead of silently falling back to `number`.

File: apps/channels/forms.py [hunk 2]
```diff
@@ -235,14 +235,33 @@ def clean_number(self):
...
+                except Exception:
+                    raise forms.ValidationError(
+                        "Could not validate this number with Meta right now. Please try again."
+                    ) from None
+                if not phone_number_id:
+                    raise forms.ValidationError(
+                        f"{number} was not found in the WhatsApp Business Account. "
+                        "Please verify the number is registered with your business."
+                    )
+                self._phone_number_id = phone_number_id
```
Comment (snopoke) [inline apps/channels/forms.py]: This feels hacky - what about doing basic phone number validation here (make sure it's a correctly formatted phone number) and then moving the service level check to the `clean` method and putting the `phone_number_id` into `form.cleaned_data`.

I think that would follow normal Django patterns better and you won't need to rely on a temporary field on the form.

File: apps/channels/meta_webhook.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    return HttpResponseBadRequest("Verification failed.")
+
+
+def verify_signature(payload: bytes, signature_header: str, app_secret: str) -> bool:
+    """Verify the X-Hub-Signature-256 header from Meta webhooks."""
+    if not signature_header.startswith("sha256=") or not app_secret:
+        return False
+
+    expected_signature = signature_header[7:]
```
Comment (snopoke) [inline apps/channels/meta_webhook.py]: prefer `removeprefix`:
```suggestion
    expected_signature = signature_header.removeprefix("sha256=")
```

File: apps/channels/forms.py [hunk 4]
```diff
@@ -235,14 +235,33 @@ def clean_number(self):
             number_obj = phonenumbers.parse(self.cleaned_data["number"])
             number = phonenumbers.format_number(number_obj, phonenumbers.PhoneNumberFormat.E164)
             service = self.messaging_provider.get_messaging_service()
-            if not service.is_valid_number(number):
+            if self.messaging_provider.type == MessagingProviderType.meta_cloud_api:
```
[the remark this answers is not in the document]
Comment (SmittieC) [reply apps/channels/forms.py]: Fixed in 4a5ca3970. Added `is_valid_number` to `MetaCloudAPIService` which wraps the phone number ID lookup. The form's `clean` method now calls `service.is_valid_number(number)` uniformly for all providers — no provider-type check needed for validation.

File: apps/channels/forms.py (hunk 2 above)
Comment (SmittieC) [reply apps/channels/forms.py]: Fixed in 4a5ca3970. `clean_number` now only does phone number format validation. The service-level check (`is_valid_number`) moved to `clean`, and `phone_number_id` goes into `cleaned_data` directly — no more temp attribute or `post_save` override.

File: apps/channels/views.py [hunk 5]
```diff
@@ -418,3 +419,74 @@ def delete_channel(request, team_slug, experiment_id: int, channel_id: int):
...
+
+        # A single Meta webhook payload is always scoped to one app, so all entries share
+        # the same app_secret. Verifying with the first channel's secret is sufficient.
+        log.debug("Meta Cloud API webhook dispatching %d message(s) for channel %s", len(message_values), channel.id)
+        for value in message_values:
+            tasks.handle_meta_cloud_api_message.delay(
+                channel_id=channel.id,
+                team_slug=channel.team.slug,
+                message_data=value,
+            )
```
[the remark this answers is not in the document]
Comment (snopoke) [reply apps/channels/views.py]: @SmittieC is this a real risk?

[the remark this answers is not in the document]
Comment (SmittieC) [reply apps/channels/views.py]: Yes, but an unlikely one. Still needs fixing though

File: apps/channels/tasks.py [hunk 6]
```diff
@@ -202,6 +203,7 @@ def handle_meta_cloud_api_message(self, channel_id: int, team_slug: str, message
         .select_related("experiment", "team", "messaging_provider")
         .first()
     )
+    set_current_team(experiment_channel.team)
```
Comment (snopoke) [inline apps/channels/tasks.py]: this should be safe to do because we're using the thread pool but it would probably be a good idea to add a post task signal to unset the team context as well (see `config/celery.py`)

(separate PR is fine)
